Move from failure to thiserror

poros opened the pull request:
Closes #115

This is still a WIP branch, with lots of TODOs and some things about
thiserror I still can't wrap my head around. However, the heavy-lifting
is done, the failure crate has been removed from the list of
dependencies and compilation, tests, benchmarks and linting are all
green.

The two biggest things I have yet to figure out are:
1. How to deal with the errors manually defined in ser.rs and de.rs:
   they are publicly available and as soon as I touch anything I hit
   cryptic serde errors
2. How to make errors like TryFromIntError part of more abstract ones
   like ParseSchemaError, which could have a source error or just a
   String description.

poros commented:
@cpcloud and @athre0z: I would really appreciate your early feedback, if you have any, as you both commented on the original issue and you are likely more familiar with error handling in rust than me.

cpcloud commented on src/errors.rs:
```diff
@@ -0,0 +1,73 @@
+
+#[derive(Error, Debug)]
+/// Error type returned from the library
+pub enum AvroError {
```
I think it makes sense to call this `Error`, since it's already in the `avro_rs` crate. If users want to rename it the option is always available when importing.

poros replied:
Yes, you are right. I started with `AvroError` to makes my `sed` easier and then I kinda rolled with it, but `Error` is the best choice.

cpcloud commented on src/errors.rs:
```diff
@@ -0,0 +1,73 @@
+pub enum AvroError {
+    /// Represents all cases of `std::io::Error`
+    #[error(transparent)]
+    IOError(#[from] std::io::Error),
```
At my company we've been adopting a slightly different convention for naming error variants, and that naming convention is simply to remove the redundant `Error` suffix on the variants. It may seem awkward at first, but after awhile it grew on me, and now I prefer this way of writing errors. Since you've got a few catch-all variants in here, it'll look weird, but after converting the variants to reflect specific failure modes it's a lot better. So you end up with error constructions like: `Error::Decode`, `Error::ParseSchema`, etc.

cpcloud replied:
I don't consider this a blocker for landing the PR.

athre0z replied:
Love the idea, I'll absolutely adapt that for my stuff!

poros replied:
Ah, I think this is in line with the `thiserror` documentation, too. I kinda copied most of the error names (made bulk of changes easier), but I like this approach better!

cpcloud commented on src/errors.rs:
```diff
@@ -0,0 +1,73 @@
+    #[error(transparent)]
+    JSONError(#[from] serde_json::Error),
+
+    // TODO: Should this be wrapped by SchemaResolutionError somehow?
```
Just an observation but I think that many of these `TODO`s go away if you make more specific error variants.

poros replied:
I think you are right. I think I missed the fact that having many errors is not a problem in Rust. Although, I am curious to see at what point I should go for the approach that @athre0z suggested above:
>If you have many many failure cases I'd even recommend splitting them up into multiple different error types, e.g. DecodingError, NetworkError that then can all be into'ed a common error enum type with branches for each more granular error type. This allows for being more verbose in describing what exactly went wrong without forcing users to match on an unreasonable amount of error types when they actually just want to catch e.g. a decoding error in general.

athre0z commented on src/errors.rs:
```diff
@@ -0,0 +1,73 @@
+use crate::ser;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
```
I'd probably add a `#[non_exhaustive]` here to ensure that adding new error variants remains a non-API-breaking operation. It's available since Rust version 1.40.

poros replied:
Very good suggestion, thanks!

poros commented:
Thank you both @cpcloud and @athre0z ! This was some fantastic and blazing fast feedback! You are awesome ❤️ 

I'll try to respond inline and to incorporate everything as soon as I get again some time to dedicate to this.

poros commented on tests/io.rs:
```diff
@@ -220,10 +218,13 @@ fn test_no_default_value() -> Result<(), String> {
-            Some(_) => Ok(()),
-            None => Err(format!("Expected SchemaResolutionError, got {}", e)),
-        },
+        Err(ref e) => {
+            if let AvroError::SchemaResolutionError(_) = e {
+                Ok(())
+            } else {
+                Err(format!("Expected SchemaResolutionError, got {}", e))
+            }
+        }
```
Ah, good catch! I became dull by editing mechanically so many lines :P

poros commented on src/codec.rs:
```diff
@@ -40,27 +39,28 @@ impl ToAvro for Codec {
             #[cfg(feature = "snappy")]
             "snappy" => Ok(Codec::Snappy),
-            _ => Err(DecodeError::new("unrecognized codec")),
+            _ => Err(AvroError::DecodeError("unrecognized codec".to_string())),
```
I am happy to hear that you'd like to contribute to writing error variants. I honestly don't know how many of them is too many or too few, so a second opinion would definitely be appreciated.

I am happy to merge this one with a few more of the suggested changes so that you can do a second pass or to merge your PR into my fork. Feel free to choose what works best for you.

poros commented on src/codec.rs:
```diff
@@ -104,7 +104,7 @@ impl Codec {
 
                 if expected_crc != actual_crc {
-                    return Err(DecodeError::new(format!(
+                    return Err(AvroError::DecodeError(format!(
```
Gotcha, thanks!

poros commented:
> Personally, I tend create a custom result alias such as `type AvroResult<T> = Result<T, AvroError>` for my crates to spare me some typing, but that's just a personal preference.

That's a good suggestion. I'll do that.

> You can just impl From<TryFromIntError> for AvroError manually for such cases.
Ah, right. That was much more straightforward than I thought!

> I'm not really too much of a serde expert, so I can't really comment on that specifically. Generally speaking, I don't consider having more than one error type per crate an issue, as long as the error types can be into()ed the main error type and isn't type erased. This is already ensured thanks to you using #[from] in the corresponding AvroError enum branch.
Perhaps you are right, I could probably just let them be.

> If you have many many failure cases I'd even recommend splitting them up into multiple different error types, e.g. DecodingError, NetworkError that then can all be into'ed a common error enum type with branches for each more granular error type. This allows for being more verbose in describing what exactly went wrong without forcing users to match on an unreasonable amount of error types when they actually just want to catch e.g. a decoding error in general.
Ah, this is a bit of a mix between one of the recipes that `failure` had and what seems to be the general way to use `thiserror`. It's interesting, but I don't know how many variants is enough to justify the introduction of such pattern (I suppose that users would find a bit more complicated to pattern match on the variants of the sub-errors)

cpcloud commented:
My 2c: I think that it's best to have a single error variant for each specific error case, even if it means a gigantic enum. Users of the library that are other libraries using `thiserror` will be able to use `SomeVariant(#[source] avro_rs::Error)` to bubble errors up, unless they want to handle a specific error. Users of `avro_rs` that are executables should be using `anyhow`, in which case they can use the `?` operator (possibly `with_context`) and that's enough to get nice error tracebacks.

For example, even if you're `read_exact`ing in a number of different places, each `read_exact` call is likely for a different purpose, e.g. reading a double, reading a float, etc. As a user of the library I want to know exactly what was happening when an error is triggered (so, `Error::ReadDouble`, `Error::ReadFloat`, etc). I don't want to get a non-specific `Error::IOError` that merely wraps a `std::io::Error`, otherwise why not return `io::Error` in those places?

I think this means that having multiple error types only serves to make it more convenient to type for developers of `avro_rs` and I'm not sure that's actually something to optimize at the moment :)

poros commented:
@cpcloud , @athre0z : I feel that I have addressed most of the comments and suggestions. I believe that as a first this PR might be pretty decent, but there is plenty of room for improvement. It would be great if you could look at it again :)

poros commented:
Thanks to you both! :)
